test/integration/sds_dynamic_integration_test.cc

```diff
@@ -974,5 +978,168 @@ TEST_P(SdsCdsIntegrationTest, BasicSuccess) {
+    (*val.mutable_fields())["private_key_file"].set_string_value(
+        TestEnvironment::temporaryPath("root/current/serverkey.pem"));
+    (*val.mutable_fields())["expected_operation"].set_string_value("sign");
+    (*val.mutable_fields())["sync_mode"].set_bool_value(true);
+
+    // check the key type and set the mode accordingly
+    std::string mode;
+    std::string private_key = TestEnvironment::readFileToStringForTest(
+        TestEnvironment::temporaryPath("root/current/serverkey.pem"));
+    bssl::UniquePtr<BIO> bio(
+        BIO_new_mem_buf(const_cast<char*>(private_key.data()), private_key.size()));
+    bssl::UniquePtr<EVP_PKEY> pkey(PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr));
+
+    EXPECT_TRUE(pkey != nullptr);
+
+    if (EVP_PKEY_id(pkey.get()) == EVP_PKEY_RSA) {
+      mode = "rsa";
+    } else if (EVP_PKEY_id(pkey.get()) == EVP_PKEY_EC) {
+      mode = "ecdsa";
+    }
+    (*val.mutable_fields())["mode"].set_string_value(mode);
+
+    secret.set_name(client_cert_);
+    auto* tls_certificate = secret.mutable_tls_certificate();
+    tls_certificate->mutable_certificate_chain()->set_filename(
+        TestEnvironment::temporaryPath("root/current/servercert.pem"));
+    tls_certificate->mutable_private_key_provider()->set_provider_name("test");
+    tls_certificate->mutable_private_key_provider()->mutable_typed_config()->set_type_url(
+        "type.googleapis.com/google.protobuf.Struct");
+    tls_certificate->mutable_private_key_provider()->mutable_typed_config()->PackFrom(val);
+    auto* watched_directory = tls_certificate->mutable_watched_directory();
+    watched_directory->set_path(TestEnvironment::temporaryPath("root"));
+    return secret;
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersionsClientType, SdsCdsPrivateKeyIntegrationTest,
+                         testing::ValuesIn(getSdsTestsParams(true)), sdsTestParamsToString);
+
+TEST_P(SdsCdsPrivateKeyIntegrationTest, BasicSdsCdsPrivateKeyProvider) {
+  v3_resource_api_ = true;
+
+  TestEnvironment::exec(
+      {TestEnvironment::runfilesPath("test/integration/sds_dynamic_key_rotation_setup.sh")});
+
+  // Set up the private key provider.
+  Extensions::PrivateKeyMethodProvider::TestPrivateKeyMethodFactory test_factory;
+  Registry::InjectFactory<Ssl::PrivateKeyMethodProviderInstanceFactory>
+      test_private_key_method_factory(test_factory);
+
+  on_server_init_function_ = [this]() {
+    {
+      // CDS.
+      AssertionResult result =
+          fake_upstreams_[1]->waitForHttpConnection(*dispatcher_, xds_connection_);
+      RELEASE_ASSERT(result, result.message());
+      result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+      RELEASE_ASSERT(result, result.message());
+      xds_stream_->startGrpcStream();
+      sendCdsResponse();
+    }
+    {
+      // SDS.
+      AssertionResult result =
+          fake_upstreams_[2]->waitForHttpConnection(*dispatcher_, sds_connection_);
+      RELEASE_ASSERT(result, result.message());
```

Comment:
Why are these `RELEASE_ASSERT` in tests rather than `ASSERT_TRUE`, `EXPECT_TRUE` etc. gtest predicates?

Reply:
That seems to be the style of this file: `AssertionResult` values are checked with `RELEASE_ASSERT`. @dio I see you have suggested this at least in https://github.com/envoyproxy/envoy-filter-example/issues/56#issuecomment-418689717, can you comment?

Reply:
This doesn't seem right, I think we see more informative error messages when using gtest assertion macros, e.g. see https://github.com/google/googletest/blob/master/docs/advanced.md#predicate-assertions-for-better-error-messages.